File: ios/Classes/ProVideoEditorPlugin.swift
import Flutter
import UIKit

public class ProVideoEditorPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {
  var eventSink: FlutterEventSink?
  // Buffer last known progress by id so that listeners that attach later can receive the most
  // recent update instead of starting at 0.0.
  // Store progress and optional stage for buffered events
  var lastProgress: [String: (progress: Double, stage: String?)] = [:]

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(
      name: "pro_video_editor", binaryMessenger: registrar.messenger())
    let eventChannel = FlutterEventChannel(
      name: "pro_video_editor_progress", binaryMessenger: registrar.messenger())

    let instance = ProVideoEditorPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)
    eventChannel.setStreamHandler(instance)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)

    case "getMetadata":
      guard let args = call.arguments as? [String: Any],
        let inputPath = args["inputPath"] as? String,
        let extensionStr = args["extension"] as? String
      else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENTS", message: "Expected arguments missing", details: nil))
        return
      }

      Task {
        do {
          let meta = try await VideoMetadata.processVideo(inputPath: inputPath, ext: extensionStr)
          result(meta)
        } catch {
          result(
            FlutterError(code: "METADATA_ERROR", message: error.localizedDescription, details: nil))
        }
      }

    case "getThumbnails":
      guard let args = call.arguments as? [String: Any],
        let id = args["id"] as? String,
        let inputPath = args["inputPath"] as? String,
        let extensionStr = args["extension"] as? String,
        let boxFit = args["boxFit"] as? String,
        let outputFormat = args["outputFormat"] as? String,
        let outputWidth = args["outputWidth"] as? Int,
        let outputHeight = args["outputHeight"] as? Int
      else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing parameters", details: nil))
        return
      }

      let timestampsUs = (args["timestamps"] as? [NSNumber])?.map { $0.int64Value } ?? []
      let maxOutputFrames = args["maxOutputFrames"] as? Int

      postProgress(id: id, progress: 0.0)

      Task {
        let thumbnails = await ThumbnailGenerator.getThumbnails(
          inputPath: inputPath,
          extension: extensionStr,
          outputFormat: outputFormat,
          boxFit: boxFit,
          outputWidth: outputWidth,
          outputHeight: outputHeight,
          timestampsUs: timestampsUs,
          maxOutputFrames: maxOutputFrames,
          onProgress: { progress, stage in
            self.postProgress(id: id, progress: progress, stage: stage)
          }
        )
        self.postProgress(id: id, progress: 1.0)
        result(thumbnails)
      }

    case "renderVideo":
      guard let args = call.arguments as? [String: Any],
        let id = args["id"] as? String,
        let inputPath = args["inputPath"] as? String
      else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENTS", message: "Missing parameters", details: nil))
        return
      }
      
      let inputFormat = args["inputFormat"] as? String ?? "mp4"
      let outputFormat = args["outputFormat"] as? String ?? "mp4"
      let outputPath = args["outputPath"] as? String
      let imageBytes = (args["imageBytes"] as? FlutterStandardTypedData)?.data
      let rotateTurns = args["rotateTurns"] as? Int
      let cropWidth = args["cropWidth"] as? Int
      let cropHeight = args["cropHeight"] as? Int
      let cropX = args["cropX"] as? Int
      let cropY = args["cropY"] as? Int
      let scaleX = (args["scaleX"] as? NSNumber)?.floatValue
      let scaleY = (args["scaleY"] as? NSNumber)?.floatValue
      let flipX = args["flipX"] as? Bool ?? false
      let flipY = args["flipY"] as? Bool ?? false
      let blur = args["blur"] as? Double
      let bitrate = args["bitrate"] as? Int
      let enableAudio = args["enableAudio"] as? Bool ?? true
      let playbackSpeed = (args["playbackSpeed"] as? NSNumber)?.floatValue
      let startUs = args["startTime"] as? Int64
      let endUs = args["endTime"] as? Int64
      let colorMatrixList = args["colorMatrixList"] as? [[Double]] ?? []
      
      // Custom audio parameters
      let customAudioPath = args["customAudioPath"] as? String
      let customAudioVolume = (args["customAudioVolume"] as? NSNumber)?.doubleValue ?? 1.0
      let customAudioStartTime = args["customAudioStartTime"] as? Int64
      let customAudioEndTime = args["customAudioEndTime"] as? Int64
      let customAudioFadeInDuration = args["customAudioFadeInDuration"] as? Int64 ?? 0
      let customAudioFadeOutDuration = args["customAudioFadeOutDuration"] as? Int64 ?? 0

      postProgress(id: id, progress: 0.0)

      let preferH264 = args["preferH264"] as? Bool ?? false

      RenderVideo.render(
        id: id,
        inputPath: inputPath,
        imageData: imageBytes,
        inputFormat: inputFormat,
        outputFormat: outputFormat,
        outputPath: outputPath,
        rotateTurns: rotateTurns,
        flipX: flipX,
        flipY: flipY,
        cropWidth: cropWidth,
        cropHeight: cropHeight,
        cropX: cropX,
        cropY: cropY,
        scaleX: scaleX,
        scaleY: scaleY,
        bitrate: bitrate,
        enableAudio: enableAudio,
        playbackSpeed: playbackSpeed,
        startUs: startUs,
        endUs: endUs,
        colorMatrixList: colorMatrixList,
        blur: blur,
        customAudioPath: customAudioPath,
        customAudioVolume: customAudioVolume,
        customAudioStartTime: customAudioStartTime,
        customAudioEndTime: customAudioEndTime,
        customAudioFadeInDuration: customAudioFadeInDuration,
        customAudioFadeOutDuration: customAudioFadeOutDuration,
        preferH264: preferH264,
        onProgress: { progress, stage in
          self.postProgress(id: id, progress: progress, stage: stage)
        },
        onComplete: { outputData in
          self.postProgress(id: id, progress: 1.0)
          result(outputData)
        },
        onError: { error in
          result(
            FlutterError(code: "RENDER_ERROR", message: error.localizedDescription, details: nil))
        }
      )

    case "concatenateVideos":
      guard let args = call.arguments as? [String: Any],
        let id = args["id"] as? String,
        let inputPaths = args["inputPaths"] as? [String],
        let outputPath = args["outputPath"] as? String
      else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENTS", message: "Missing parameters for concatenateVideos", details: nil))
        return
      }
      
      postProgress(id: id, progress: 0.0)
      
      ConcatenateVideos.concatenate(
        inputPaths: inputPaths,
        outputPath: outputPath,
        onProgress: { progress, stage in
          self.postProgress(id: id, progress: progress, stage: stage)
        },
        onComplete: { outputPath in
          self.postProgress(id: id, progress: 1.0)
          result(outputPath)
        },
        onError: { error in
          result(
            FlutterError(code: "CONCATENATE_ERROR", message: error.localizedDescription, details: nil))
        }
      )

    case "generateSlideshow":
      guard let args = call.arguments as? [String: Any],
        let id = args["id"] as? String,
        let slidesData = args["slides"] as? [[String: Any]],
        let outputPath = args["outputPath"] as? String
      else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENTS", message: "Missing parameters for generateSlideshow", details: nil))
        return
      }
      
      let width = args["width"] as? Int ?? 1920
      let height = args["height"] as? Int ?? 1080
      let fps = args["fps"] as? Int ?? 30
      let audioPath = args["audioPath"] as? String
      
      // Audio parameters
      let audioTrimStartMs = (args["audioTrimStartMs"] as? NSNumber)?.int64Value
      let audioTrimEndMs = (args["audioTrimEndMs"] as? NSNumber)?.int64Value
      let audioVolume = (args["audioVolume"] as? NSNumber)?.doubleValue
      let audioFadeInMs = (args["audioFadeInMs"] as? NSNumber)?.int64Value
      let audioFadeOutMs = (args["audioFadeOutMs"] as? NSNumber)?.int64Value
      
      // Parse slides data
      let slides: [SlideshowGenerator.SlideConfig] = slidesData.compactMap { slideMap in
        guard let imagePath = slideMap["imagePath"] as? String,
              let durationMs = (slideMap["durationMs"] as? NSNumber)?.int64Value
        else {
          return nil
        }
        
        let transitionInType = slideMap["transitionInType"] as? String ?? "fade"
        let transitionOutType = slideMap["transitionOutType"] as? String ?? "fade"
        let transitionInDurationMs = (slideMap["transitionInDurationMs"] as? NSNumber)?.int64Value ?? 500
        let transitionOutDurationMs = (slideMap["transitionOutDurationMs"] as? NSNumber)?.int64Value ?? 500
        
        return SlideshowGenerator.SlideConfig(
          imagePath: imagePath,
          durationMs: durationMs,
          transitionInType: transitionInType,
          transitionOutType: transitionOutType,
          transitionInDurationMs: transitionInDurationMs,
          transitionOutDurationMs: transitionOutDurationMs
        )
      }
      
      guard !slides.isEmpty else {
        result(
          FlutterError(
            code: "INVALID_ARGUMENTS", message: "No valid slides provided", details: nil))
        return
      }
      
      postProgress(id: id, progress: 0.0)
      
      SlideshowGenerator.generateSlideshow(
        slides: slides,
        outputPath: outputPath,
        width: width,
        height: height,
        fps: fps,
        audioPath: audioPath,
        audioTrimStartMs: audioTrimStartMs,
        audioTrimEndMs: audioTrimEndMs,
        audioVolume: audioVolume,
        audioFadeInMs: audioFadeInMs,
        audioFadeOutMs: audioFadeOutMs,
        onProgress: { progress, stage in
          self.postProgress(id: id, progress: progress, stage: stage)
        },
        onComplete: { resultPath in
          self.postProgress(id: id, progress: 1.0)
          result(resultPath)
        },
        onError: { error in
          result(
            FlutterError(code: "SLIDESHOW_ERROR", message: error.localizedDescription, details: nil))
        }
      )

    case "cancelRender":
      guard let args = call.arguments as? [String: Any], let id = args["id"] as? String else {
        result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing id", details: nil))
        return
      }
      // Attempt to cancel and return result
      let cancelled = RenderVideo.cancel(id: id)
      // Clear any buffered progress for this id to avoid stale events
      lastProgress.removeValue(forKey: id)
      result(cancelled)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  private func postProgress(id: String, progress: Double, stage: String? = nil) {
    DispatchQueue.main.async {
      // store buffered value
      self.lastProgress[id] = (progress: progress, stage: stage)
      print("ProVideoEditorPlugin.postProgress(\(id)) -> \(progress) sink=\(self.eventSink != nil)")
      self.eventSink?([
        "id": id,
        "progress": progress,
        "stage": stage as Any,
      ])
      if progress >= 1.0 {
        self.lastProgress.removeValue(forKey: id)
      }
    }
  }

  @objc public func onListen(
    withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink
  ) -> FlutterError? {
    self.eventSink = events
    print("ProVideoEditorPlugin.onListen - sink set")
    // Emit buffered progress values for all tasks so UI has a fast snapshot
    for (id, p) in lastProgress {
      self.eventSink?([
        "id": id,
        "progress": p.progress,
        "stage": p.stage as Any,
      ])
    }
    return nil
  }

  @objc public func onCancel(withArguments arguments: Any?) -> FlutterError? {
    print("ProVideoEditorPlugin.onCancel - sink cleared")
    self.eventSink = nil
    return nil
  }
}

File: android/src/main/kotlin/ch/waio/pro_video_editor/ProVideoEditorPlugin.kt
package ch.waio.pro_video_editor

import android.os.Handler
import android.os.Looper
import android.util.Log
import ch.waio.pro_video_editor.src.features.ConcatenateVideos
import ch.waio.pro_video_editor.src.features.Metadata
import ch.waio.pro_video_editor.src.features.render.RenderVideo
import ch.waio.pro_video_editor.src.features.SlideshowGenerator
import ch.waio.pro_video_editor.src.features.ThumbnailGenerator
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import kotlinx.coroutines.*

/** ProVideoEditorPlugin */
class ProVideoEditorPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var methodChannel: MethodChannel
    private lateinit var eventChannel: EventChannel
    private var eventSink: EventChannel.EventSink? = null
    // Buffer last known progress by id so we don't lose updates when sink reconnects
    // Store both progress value and optional stage string so listeners can
    // distinguish between rendering vs mixing and show appropriate UI.
    private data class ProgressEntry(val progress: Double, val stage: String?)
    private val lastProgress = java.util.concurrent.ConcurrentHashMap<String, ProgressEntry>()

    private lateinit var renderVideo: RenderVideo
    private lateinit var metadata: Metadata
    private lateinit var thumbnailGenerator: ThumbnailGenerator
    private lateinit var concatenateVideos: ConcatenateVideos
    private lateinit var slideshowGenerator: SlideshowGenerator

    private val coroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel = MethodChannel(flutterPluginBinding.binaryMessenger, "pro_video_editor")
        eventChannel =
            EventChannel(flutterPluginBinding.binaryMessenger, "pro_video_editor_progress")

        methodChannel.setMethodCallHandler(this)
        eventChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                eventSink = events
                Log.d(PACKAGE_TAG, "EventChannel onListen - sink set: ${eventSink != null}")
                // Emit any buffered progress values to new listener
                // We want to send the most recent progress for each task
                lastProgress.forEach { (id, p) ->
                    try {
                        eventSink?.success(mapOf("id" to id, "progress" to p.progress, "stage" to p.stage))
                    } catch (e: Exception) {
                        Log.w(PACKAGE_TAG, "Failed to dispatch buffered progress for $id: ${e.message}")
                    }
                }
            }

            override fun onCancel(arguments: Any?) {
                Log.d(PACKAGE_TAG, "EventChannel onCancel - sink cleared")
                eventSink = null
            }
        })

        renderVideo = RenderVideo(flutterPluginBinding.applicationContext);
        metadata = Metadata(flutterPluginBinding.applicationContext)
        thumbnailGenerator = ThumbnailGenerator(flutterPluginBinding.applicationContext)
        concatenateVideos = ConcatenateVideos(flutterPluginBinding.applicationContext)
        slideshowGenerator = SlideshowGenerator(flutterPluginBinding.applicationContext)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            "getMetadata" -> {
                val inputPath = call.argument<String>("inputPath") 
                val extension = call.argument<String>("extension")

                if (inputPath != null && extension != null) {
                    val meta = metadata.processVideo(inputPath, extension)
                    result.success(meta)
                } else {
                    result.error(
                        "InvalidArgument", "Expected raw Uint8List (ByteArray/List<Int>)", null
                    )
                }
            }

            "getThumbnails" -> {
                val id = call.argument<String>("id") ?: ""
                val inputPath = call.argument<String>("inputPath") 
                val extension = call.argument<String>("extension")
                val boxFit = call.argument<String>("boxFit")
                val outputFormat = call.argument<String>("outputFormat")
                val outputWidth = call.argument<Number>("outputWidth")?.toInt()
                val outputHeight = call.argument<Number>("outputHeight")?.toInt()
                val rawTimestamps = call.argument<List<Number>>("timestamps") ?: emptyList()
                val timestampsUs = rawTimestamps.map { it.toLong() }
                val maxOutputFrames = call.argument<Number>("maxOutputFrames")?.toInt()


                if (inputPath == null ||
                    extension == null ||
                    boxFit == null ||
                    outputFormat == null ||
                    outputWidth == null ||
                    outputHeight == null ||
                    (timestampsUs == null && maxOutputFrames == null)
                ) {
                    result.error("INVALID_ARGUMENTS", "Missing or invalid arguments", null)
                    return
                }
                postProgress(id, 0.0)

                coroutineScope.launch {
                    try {
                        val thumbnails = thumbnailGenerator.getThumbnails(
                            inputPath = inputPath,
                            extension = extension,
                            outputFormat = outputFormat,
                            boxFit = boxFit,
                            outputWidth = outputWidth,
                            outputHeight = outputHeight,
                            timestampsUs = timestampsUs,
                            maxOutputFrames = maxOutputFrames,
                                onProgress = { progress, stage -> postProgress(id, progress, stage) },
                        )

                        withContext(Dispatchers.Main) {
                            postProgress(id, 1.0)
                            result.success(thumbnails)
                        }
                    } catch (e: Exception) {
                        withContext(Dispatchers.Main) {
                            result.error("THUMBNAIL_ERROR", e.message, null)
                        }
                    }
                }
            }

            "renderVideo" -> {
                val id = call.argument<String>("id") ?: ""
                val imageBytes = call.argument<ByteArray?>("imageBytes")
                val rotateTurns = call.argument<Number>("rotateTurns")?.toInt()
                val cropWidth = call.argument<Number>("cropWidth")?.toInt()
                val cropHeight = call.argument<Number>("cropHeight")?.toInt()
                val cropX = call.argument<Number>("cropX")?.toInt()
                val cropY = call.argument<Number>("cropY")?.toInt()
                val bitrate = call.argument<Number>("bitrate")?.toInt()
                val scaleX = call.argument<Number>("scaleX")?.toFloat()
                val scaleY = call.argument<Number>("scaleY")?.toFloat()
                val blur = call.argument<Number>("blur")?.toDouble()
                val flipX = call.argument<Boolean>("flipX") ?: false
                val flipY = call.argument<Boolean>("flipY") ?: false
                val enableAudio = call.argument<Boolean>("enableAudio") ?: true
                val playbackSpeed = call.argument<Number>("playbackSpeed")?.toFloat()
                val startUs = call.argument<Number>("startTime")?.toLong()
                val endUs = call.argument<Number>("endTime")?.toLong()
                val inputFormat = call.argument<String>("inputFormat") ?: "mp4"
                val outputFormat = call.argument<String>("outputFormat") ?: "mp4"
                val inputPath = call.argument<String>("inputPath") ?: ""
                val outputPath = call.argument<String>("outputPath")
                val colorMatrixList = call.argument<List<List<Double>>>("colorMatrixList")
                    ?: emptyList<List<Double>>()
                
                // Custom audio parameters
                val customAudioPath = call.argument<String?>("customAudioPath")
                val customAudioVolume = call.argument<Number>("customAudioVolume")?.toDouble() ?: 1.0
                val customAudioStartTime = call.argument<Number>("customAudioStartTime")?.toLong()
                val customAudioEndTime = call.argument<Number>("customAudioEndTime")?.toLong()
                val customAudioFadeInDuration = call.argument<Number>("customAudioFadeInDuration")?.toLong() ?: 0L
                val customAudioFadeOutDuration = call.argument<Number>("customAudioFadeOutDuration")?.toLong() ?: 0L

                postProgress(id, 0.0)

                renderVideo.render(
                    id = id,
                    imageBytes = imageBytes,
                    inputFormat = inputFormat,
                    outputFormat = outputFormat,
                    inputPath = inputPath,
                    outputPath = outputPath,
                    rotateTurns = rotateTurns,
                    flipX = flipX,
                    flipY = flipY,
                    scaleX = scaleX,
                    scaleY = scaleY,
                    cropWidth = cropWidth,
                    cropHeight = cropHeight,
                    cropX = cropX,
                    cropY = cropY,
                    enableAudio = enableAudio,
                    playbackSpeed = playbackSpeed,
                    startUs = startUs,
                    endUs = endUs,
                    colorMatrixList = colorMatrixList,
                    blur = blur,
                    bitrate = bitrate,
                    customAudioPath = customAudioPath,
                    customAudioVolume = customAudioVolume,
                    customAudioStartTime = customAudioStartTime,
                    customAudioEndTime = customAudioEndTime,
                    customAudioFadeInDuration = customAudioFadeInDuration,
                    customAudioFadeOutDuration = customAudioFadeOutDuration,
                    onProgress = { progress, stage -> postProgress(id, progress, stage) },
                    onComplete = { resultBytes ->
                        postProgress(id, 1.0)
                        Handler(Looper.getMainLooper()).post {
                            result.success(resultBytes)
                        }
                    },
                    onError = { error ->
                        Log.e("RenderVideo", "Error rendering video: ${error.message}")
                    }
                )
            }

            "concatenateVideos" -> {
                val id = call.argument<String>("id") ?: ""
                val inputPaths = call.argument<List<String>>("inputPaths")
                val outputPath = call.argument<String>("outputPath")

                if (inputPaths == null || inputPaths.isEmpty() || outputPath == null) {
                    result.error("INVALID_ARGUMENTS", "Missing or invalid arguments", null)
                    return
                }

                postProgress(id, 0.0)

                coroutineScope.launch {
                    concatenateVideos.concatenate(
                        inputPaths = inputPaths,
                        outputPath = outputPath,
                            onProgress = { progress, stage -> postProgress(id, progress, stage) },
                        onComplete = { resultPath ->
                            Handler(Looper.getMainLooper()).post {
                                postProgress(id, 1.0)
                                result.success(resultPath)
                            }
                        },
                        onError = { error ->
                            Handler(Looper.getMainLooper()).post {
                                result.error("CONCATENATE_ERROR", error.message, null)
                            }
                        }
                    )
                }
            }

            "generateSlideshow" -> {
                val id = call.argument<String>("id") ?: ""
                val slidesData = call.argument<List<Map<String, Any>>>("slides")
                val outputPath = call.argument<String>("outputPath")
                val width = call.argument<Number>("width")?.toInt() ?: 1920
                val height = call.argument<Number>("height")?.toInt() ?: 1080
                val fps = call.argument<Number>("fps")?.toInt() ?: 30
                val audioPath = call.argument<String?>("audioPath")
                
                // Audio parameters
                val audioTrimStartMs = call.argument<Number?>("audioTrimStartMs")?.toLong()
                val audioTrimEndMs = call.argument<Number?>("audioTrimEndMs")?.toLong()
                val audioVolume = call.argument<Number?>("audioVolume")?.toDouble()
                val audioFadeInMs = call.argument<Number?>("audioFadeInMs")?.toLong()
                val audioFadeOutMs = call.argument<Number?>("audioFadeOutMs")?.toLong()

                if (slidesData == null || slidesData.isEmpty() || outputPath == null) {
                    result.error("INVALID_ARGUMENTS", "Missing or invalid arguments", null)
                    return
                }

                // Parse slides data
                val slides = slidesData.map { slideMap ->
                    SlideshowGenerator.SlideConfig(
                        imagePath = slideMap["imagePath"] as String,
                        durationMs = (slideMap["durationMs"] as Number).toLong(),
                        transitionInType = slideMap["transitionInType"] as? String ?: "fade",
                        transitionOutType = slideMap["transitionOutType"] as? String ?: "fade",
                        transitionInDurationMs = (slideMap["transitionInDurationMs"] as? Number)?.toLong() ?: 500,
                        transitionOutDurationMs = (slideMap["transitionOutDurationMs"] as? Number)?.toLong() ?: 500
                    )
                }

                postProgress(id, 0.0)

                coroutineScope.launch {
                    slideshowGenerator.generateSlideshow(
                        slides = slides,
                        outputPath = outputPath,
                        width = width,
                        height = height,
                        fps = fps,
                        audioPath = audioPath,
                        audioTrimStartMs = audioTrimStartMs,
                        audioTrimEndMs = audioTrimEndMs,
                        audioVolume = audioVolume,
                        audioFadeInMs = audioFadeInMs,
                        audioFadeOutMs = audioFadeOutMs,
                        onProgress = { progress, stage -> postProgress(id, progress, stage) },
                        onComplete = { resultPath ->
                            Handler(Looper.getMainLooper()).post {
                                postProgress(id, 1.0)
                                result.success(resultPath)
                            }
                        },
                        onError = { error ->
                            Handler(Looper.getMainLooper()).post {
                                result.error("SLIDESHOW_ERROR", error.message, null)
                            }
                        }
                    )
                }
            }

            "cancelRender" -> {
                val id = call.argument<String>("id") ?: ""
                try {
                    RenderVideo.cancelRender(id)
                    lastProgress.remove(id)
                    result.success(true)
                } catch (e: Exception) {
                    result.error("CANCEL_ERROR", e.message, null)
                }
            }

            else -> {
                result.notImplemented()
            }
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        methodChannel.setMethodCallHandler(null)
        eventChannel.setStreamHandler(null)
        eventSink = null
        coroutineScope.cancel()
    }

    private fun postProgress(id: String, progress: Double, stage: String? = null) {
        Handler(Looper.getMainLooper()).post {
            // remember last seen progress and stage
            lastProgress[id] = ProgressEntry(progress, stage)
            Log.d(PACKAGE_TAG, "postProgress($id) -> $progress (stage=${stage ?: "unknown"}), sink=${eventSink != null}")
            eventSink?.success(
                mapOf(
                    "id" to id,
                    "progress" to progress,
                    "stage" to stage
                )
            )
            // Remove buffer on completion to avoid stale events on new listeners
            if (progress >= 1.0) {
                lastProgress.remove(id)
            }
        }
    }
}
